### PR TITLE
[VIRTS-1970] Move files aside on --fresh instead of deleting them

### PR DESCRIPTION
**Pull request by bworrell**
## Description
Depends on: https://github.com/mitre/fieldmanual/pull/73

When running `server.py --fresh`, we currently delete the files under the `data` directory to reset the server state. This changes the behavior to back up the files before deleting them (creating a timestamped, gzipped tarball containing the files that are about to be deleted)

## Note
Every time you run the server with `--fresh` you will generate a new backup tarball. The backup filenames include a timestamp so you can always fine the most recent one.

Examples of directory structure pre/post `--fresh` (I didn't run `tree -a` so .gitkeep files don't appear, but they are there!):

Before:
```
data
├── abilities
│   ├── 7e422753-ad7a-4401-bc8b-b12a28e69c25.yml
│   └── testdir
│       └── 7e422753-ad7a-4401-bc8b-b12a28e69c25.yml
├── adversaries
│   └── 7e422753-ad7a-4401-bc8b-b12a28e69c25.yml
├── backup
├── object_store
├── objectives
│   └── 7e422753-ad7a-4401-bc8b-b12a28e69c25.yml
├── payloads
│   └── 7e422753-ad7a-4401-bc8b-b12a28e69c25.yml
├── results
│   └── 7e422753-ad7a-4401-bc8b-b12a28e69c25.yml
└── sources
    └── 7e422753-ad7a-4401-bc8b-b12a28e69c25.yml
```

After running w/ `--fresh`
```
data
├── abilities
├── adversaries
├── backup
│   └── backup-20210406130339.tar.gz
├── object_store
├── objectives
├── payloads
├── results
└── sources
```

Contents of the tarball
```
> tar -tzvf data/backup/backup-20210406125823.tar.gz
-rw-r--r--  0 bworrell staff     603 Apr  6 08:58 data/abilities/7e422753-ad7a-4401-bc8b-b12a28e69c25.yml
drwxr-xr-x  0 bworrell staff       0 Apr  6 08:58 data/abilities/testdir/
-rw-r--r--  0 bworrell staff     603 Apr  6 08:58 data/abilities/testdir/7e422753-ad7a-4401-bc8b-b12a28e69c25.yml
-rw-r--r--  0 bworrell staff     603 Apr  6 08:58 data/adversaries/7e422753-ad7a-4401-bc8b-b12a28e69c25.yml
-rw-r--r--  0 bworrell staff     603 Apr  6 08:58 data/objectives/7e422753-ad7a-4401-bc8b-b12a28e69c25.yml
-rw-r--r--  0 bworrell staff     603 Apr  6 08:58 data/payloads/7e422753-ad7a-4401-bc8b-b12a28e69c25.yml
-rw-r--r--  0 bworrell staff     603 Apr  6 08:58 data/results/7e422753-ad7a-4401-bc8b-b12a28e69c25.yml
-rw-r--r--  0 bworrell staff     603 Apr  6 08:58 data/sources/7e422753-ad7a-4401-bc8b-b12a28e69c25.yml
-rw-r--r--  0 bworrell staff 1137539 Apr  6 08:58 data/object_store
```

## Why?
Sometimes users run the server with `--fresh` on accident and end up accidentally deleting all their progress/work. This allows users to restore their files in this case. Note that the restore procedure is not actually documented anywhere at this time. We may choose to provide some sort of `--restore-previous-state` flag that could load the files out of the `backup` directory.

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
Still working on this. But I've run the server with `--fresh` and observed that  the tarball is created as expected. Note that this does not remove the `.gitkeep` file under each `data/*` subdirectory.

I also ran the following command to restore a  backup state:
```
> cd /path/to/caldera/repo
> tar -zxvf data/backup/backup-20210406125823.tar.gz
```

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by b17zr — APPROVED** [review]
Nit: Consider using single quotes rather than double.

**Review by b17zr — DISMISSED** [review]

File: server.py [hunk 1]
```diff
@@ -122,6 +122,7 @@ def list_str(values):
     init_swagger_documentation(app_svc.application)
 
     if args.fresh:
+        logging.info("Fresh startup: removing server data files")
```
Comment (wbooth) [inline server.py:125]: may want to add a note here on where to find previous files or just that the old session/data was saved

File: app/service/data_svc.py [hunk 2]
```diff
@@ -31,19 +43,61 @@ def __init__(self):
...
+        if os.path.exists(DATA_BACKUP_DIR):
+            DataService._delete_directory_contents(DATA_BACKUP_DIR)
+        else:
+            os.mkdir(DATA_BACKUP_DIR)
+
+        for file_path in DataService._iter_data_files():
+            src_dir, src_filename = os.path.split(file_path)
+            dst_dir = os.path.join(DATA_BACKUP_DIR, src_dir)
+            os.makedirs(dst_dir, exist_ok=True)
+            shutil.move(file_path, os.path.join(dst_dir, src_filename))
```
Comment (wbooth) [inline app/service/data_svc.py:100]: would it be easy here to compress with a timestamp in the name?

**Review by wbooth — APPROVED** [review]
not blockers but I think the updated message and zipping the files would be helpful

File: app/service/data_svc.py (hunk 2 above)
Comment (bworrell) [reply app/service/data_svc.py:100]: Yep, pretty simple. Do you think the backup directory should be cleared out every `--fresh` run, or should we allow multiple timestamped backup tarballs in there?

Comment (bworrell) [reply app/service/data_svc.py]: I changed the behavior to not clear out the backup directory and instead just always create a new, timestamped tarball

**Comment by bworrell** [comment]
I'll need to make another fieldmanual PR to update the documentation explaining the `--fresh` behavior and how to restore a backup since the approach is different now.

**Review by wbooth — APPROVED** [review]

**Review by b17zr — APPROVED** [review]

**Pull request merged**